Add validation for invalid memory references

tortmayr opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/blob/main/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

Avoid dispatching of a fetch memory request if there is not enough information available i.e no memory address is set.
Add validation for the default values of the memory address widget. => Prompts a validation error if the memory inspector has been openend for an empty/not set address 

Fixes #126

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
Start a debug session and open the memory inspector via command palette
Observe that
- no memory fetch request is sent to the debug adapter (can be verified via Memory Inspector Output)
- A validation error is displayed for the memory address field because the memory is undefined

Once the user manually changes the address, the error disappears and memory is fetched normally.

https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/assets/2311075/6a630ee6-0ae8-4a34-81f5-ca87fd958a05


#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the contribution guidelines](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/blob/main/CONTRIBUTING.md)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the code of conduct](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/blob/main/CODE_OF_CONDUCT.md)

jreineckearm commented on src/webview/memory-webview-view.tsx:
```diff
@@ -82,7 +82,7 @@ const MEMORY_DISPLAY_CONFIGURATION_DEFAULTS: MemoryDisplayConfiguration = {
 };
 
-const DEFAULT_READ_ARGUMENTS: Required<ReadMemoryArguments> = {
+export const DEFAULT_READ_ARGUMENTS: Required<ReadMemoryArguments> = {
```
Would it make sense to move this into a more common place? Like `../common/messaging` or `../common/memory`? Looks like `ReadMemoryArguments` is defined in `../common/messaging`.

tortmayr replied:
IMO I don't see any benefit in doing that.
The default constants defined in the `memory-webview-view` (MEMORY_DISPLAY_CONFIGURATION_DEFAULTS,DEFAULT_SESSION_CONTEXT etc.) only serve the purpose of constructing the initial state for the `App` component. They are an implementation detail of ( and scoped to) the webview/react app and should not leak into other components.

colin-grant-work replied:
I agree with @jreineckearm that it would be better to move this to a constants file (e.g. `view-types.ts`) to avoid circular dependencies.

jreineckearm commented:
Thanks, @tortmayr !
Handling looks good in general (see one comment about the code changes).
![image](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/assets/24247651/1aa201b1-103b-4016-aaca-2c09b1e1ff23)

Two corner cases which might be worth to address if effort is reasonable:
1. Enter only spaces in the `Address` field => `Actual` is rendered underneath the address field
![image](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/assets/24247651/9f3bc58c-0408-4045-b810-13fb41dd94d9)
2. Similar for the `Length` field (renders `Actual: 256`)
3. If you clear all fields from left to right, then change the content of the `Address` field (spaces or any value) => `Offset` and `Length` get populated with value `0`.
![image](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/assets/24247651/6dc066fc-ca85-46f8-98ba-150435823c10)

I don't know what's involved to fix item 3. It probably is the one with the least of the three findings.

tortmayr commented:
>Enter only spaces in the Address field => Actual is rendered underneath the address field

It seems like each of the three fields has its own `actual` field that is rendered if the value defined in the form field diffs from the actual set value:

![Screenshot from 2024-04-30 13-19-54](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/assets/2311075/be65bdbe-878c-4980-a7d5-cbd075e884e4)

We typically only see the actual field of the `offset` field because that's the one that naturally divergates when fetching additional rows before or after.

I guess the idea is to give the user feedback about the current value if they input an invalid value on accident.
IMO we have two options here either
1) we remove the actual fields for `address` or and `length` 
2) or we adapt the render logic and only render them if we have valid (i.e. non-empty) table data.
    Currently the options widget does not have any knowledge about the table contents so we would have to pass this info in as property. 
    
    @jreineckearm  What do you prefer?

jreineckearm commented:
Thanks, for looking into my feedback, @tortmayr !

> I guess the idea is to give the user feedback about the current value if they input an invalid value on accident.
> IMO we have two options here either
> 
>     we remove the actual fields for address or and length
> 
>     or we adapt the render logic and only render them if we have valid (i.e. non-empty) table data.
>     Currently the options widget does not have any knowledge about the table contents so we would have to pass this info in as property.

Thanks, I see now. Ideally, the options widget shouldn't know about the window status.

Revisiting this part again, I believe the label "Actual" is a bit ambiguous: it could mean both the actual value of the window contents and the last read argument. It is the latter. But let's not alter the label with this PR.

Having had a bit more of a play with it, I think what confused me was the "Actual:" label below "Address" without a value. Especially when just adding spaces to the edit field (the trimming part).

It could be enough to solve that confusion by trimming the input for comparison (and actually before passing to the memory read, I saw unexpected errors with our backend if passing in something untrimmed, though that may only be our debug adapter), and only show "Actual" below "Address" if the `memoryReference` in `activeReadArguments` is not an empty string.
Except from the initial state, `memoryReference` in `activeReadArguments` can never be the empty string because the options widget blocks an empty address field.

Does this make sense?

tortmayr commented:
@jreineckearm 
I have addressed your feedback.
- DEFAULT_READ_ARGUMENTS have been moved to common `webview-types.ts`
- Trim address value before passing to memory read. (and before validation/comparison)
- Don't show option field hints ("Actual: ...) if the webview is still in initial state

jreineckearm reviewed (approved):
Thanks, @tortmayr ! The behavior looks great now! :-)
Also thank you for moving the default arguments to a common file.

Good to go with this change
